### PR TITLE
Rpc Server alternative implementation that integrates with Microsoft's MVC

### PR DESCRIPTION
**Pull request by melanke**
This is an alternative implementation that integrates with Microsoft's MVC, it uses an ActionFilter to call the right Action based on the request and uses a Middleware to change the response.

⚠️ There is another lightweight implementation that is more flexible, easier to use, don't use a proprietary tool, has a easy-to-read source code and probably has a better performance. Check the PR: https://github.com/CityOfZion/neo-sharp/pull/243

### Code Usage
I don't know if it will be possible to add/remove a controller in runtime, maybe it will not be necessary if we use something like a permission system.

```C#
    public class HomeController : Controller
    {
        /// <summary>
        /// This method is needed, otherwise MVC will not call the filter
        /// </summary>
        public void Index()
        {}
        
        public string MethodName(int number, string message)
        {
            return $"GET /home/methodName/: {number} and {message}";
        }
    }
```

### Requests
- GET `http://localhost:10332/Home/methodName?number=2&message=hey`
- GET `http://localhost:10332/?method=methodName&params=[2, "hey"]&id=1`
- POST `http://localhost:10332/Home/methodName`
```json
{
    "number": 2,
    "message": "hey"
}
```
- POST `http://localhost:10332/`
```json
{
	"method": "methodName",
	"params": [2, "hey"]
}
```

### Response
```json
{
  "result": "GET /home/methodName/: 2 and hey"
}
```

### CODE REVIEW

**Review by osmirnov — CHANGES_REQUESTED** [review]
Please don't merge this! I am gonna try to play with it.

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcNormalizeRespMiddleware.cs
Comment (shargon) [inline src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcNormalizeRespMiddleware.cs]: We need a max-post check here (security)

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs
Comment (shargon) [inline src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs]: Max post size here too

**Review by shargon — APPROVED** [review]

**Comment by melanke** [comment]
Ok @osmirnov I will leave it in your hands 🙂

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs
Comment (shargon) [reply src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs]: @melanke Please fix this before merge

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcNormalizeRespMiddleware.cs
Comment (shargon) [reply src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcNormalizeRespMiddleware.cs]: @melanke Please fix this before merge

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs
Comment (melanke) [reply src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs]: Done

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcNormalizeRespMiddleware.cs
Comment (melanke) [reply src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcNormalizeRespMiddleware.cs]: Done

File: src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs
Comment (melanke) [reply src/NeoSharp.Core/Network/Rpc/Mvc/MvcRpcRedirectFilter.cs]: I tested and I think it's not necessary. The Middleware is executed before the Filter and stops the whole request if the post body is too big, so it doesn't get here.

**Comment by shargon** [comment]
https://docs.microsoft.com/en-us/aspnet/core/tutorials/getting-started-with-swashbuckle?view=aspnetcore-2.1&tabs=visual-studio%2Cvisual-studio-xml should be great implement this

**Comment by osmirnov** [comment]
I will explain how I tackle this but before @melanke thank you. Your the code and explanations were the great resources to understand requirements.

[I will shamelessly copy-and-tweak your comment as well]

### Requests
--------------------------------------
**Request**
- GET `http://localhost:10332/Home/methodName?number=2&message=hey`

**Output**
Controller: Home
Action: MethodName (or any method in Home marked with [ActionName("MethodName")] attribute)
Params[0].Name: number | Params[0].Value: 2 _or_ "2" | Params[0].Type: int _or_ string _or_ object
Params[1].Name: message | Params[1].Value: "hey" | Params[0].Type: string _or_ object

**Example**
```csharp
[HttpGet]
public string MethodName(int number, string message)
```

**How to get this behavior?**
This is how ASP.NET MVC default routing works

--------------------------------------
**Request**
- GET `http://localhost:10332/?method=methodName&params=[2, "hey"]&id=1`

**Output**
Controller: Home
Action: MethodName (or any method in Home marked with [ActionName("MethodName")] attribute)
Params[0].Name: params | Params[0].Value: [2, "hey"] | Params[0].Type: object[]
Params[1].Name: id | Params[1].Value: 1 _or_ "1" | Params[0].Type: int _or_ string _or_ object

**Example**
```csharp
[HttpGet]
public string MethodName([Bind(Prefix="params")] object[] parameters, string id)
```

**How to get this behavior?**
Firstly, ASP.NET MVC will suggest HomeController and Index action. We need to explain that routing has to be done based on the value of ```method``` query string parameter. Please check [MvcRpcRouter](https://github.com/CityOfZion/neo-sharp/pull/269/files#diff-c189b2687fd9a4f076f8af23246bfdb0)

Secondly, ASP.NET MVC does support params as this ([1, 2, 3]) in query string. We need to train it how to parse them. Please see [SeparatedQueryStringValueProvider](https://github.com/CityOfZion/neo-sharp/pull/269/files#diff-7989389fe55efab402b1003efbbf3ab4)

--------------------------------------
**Request**
- POST `http://localhost:10332/Home/methodName`
```json
{
    "number": 2,
    "message": "hey"
}
```

**Output**
Controller: Home
Action: MethodName (or any method in Home marked with [ActionName("MethodName")] attribute)
Params[0].Name: any | Params[0].Value: custom class instance | Params[0].Type: custom class

**Example**
```csharp
[HttpPost]
public string MethodName(NumberMessage numberMessage)
```

**How to get this behavior?**
This is how ASP.NET MVC default routing works. We only need to define the class that can be deserialized from request body as JSON

--------------------------------------
**Request**
- POST `http://localhost:10332/`
```json
{
	"method": "methodName",
	"params": [2, "hey"]
}
```

**Output**
Controller: Home
Action: MethodName (or any method in Home marked with [ActionName("MethodName")] attribute)
Params[0].Name: any | Params[0].Value: custom class instance | Params[0].Type: custom class

**Example**
```csharp
[HttpPost]
public string MethodName(MethodAndParams methodParams)
```

**How to get this behavior?**
Firstly, the same class I've used to train GET requests to be routed based on query string param is capable to read POST request body and locate method property for action name.

Secondly, ASP.NET MVC can parse [2, "hey"] and output object[] arr. I've developed simple class for this because I anticipate this is the common case. Check [MethodAndParams](https://github.com/CityOfZion/neo-sharp/pull/269/files#diff-8ac18d00e5f952c4262b719f26e35fb5)

### All classes for MVC RPC
MvcRpcServer - bootstraps MVC and all custom dependencies
MvcRpcRouter - parses action name from request query string or request body
SeparatedQueryStringValueProvider - parses array like syntax for params in query string (x=[1,"a"])
SeparatedQueryStringValueProviderFactory - injects SeparatedQueryStringValueProvider in MVC
HomeController - is default controller for short urls
MethodAndParams - is an utility class for standard POST requests
MvcRpcNormalizeRespMiddleware - ???

**Comment by osmirnov** [comment]
I don't wait it will work out-of-box because I did not even run this :D Do you guys have tests for REST?

**Comment by melanke** [comment]
@osmirnov I am using Postman to test the requests.
In your implementation, can I create a single Action Method to answer all 4 request formats?

**Comment by osmirnov** [comment]
I guess you need at least 1 method per all GET requests and 1 per all POST requests. But better to have 4 methods that we know if something does not work. You can call a single generic method inside of them. Although, I would not recommend to put all RPC logic in a single request but rather to create one controller per RESTful resource.

**Comment by osmirnov** [comment]
I will close this PR. As per discussion MVC version is viable but we want to go further and implement REST API instead of JSON-RPC. I've submitted #285 issue.

**Pull request closed without merging**
